### PR TITLE
[WIP] OAuth2: Refresh access token and retry any kind of request

### LINKED ISSUE #928: Silently refresh OAuth2 access token if expired
Default life time of access token in the OAuth2 app for ownCloud is one hour. To prevent the need of requesting the user to log-in again each time, the app needs to use the refresh token delivered with the expired access token to get a new access token silently.

Ideally this should be done whenever a network request to the server ends with an error due to unauthorized access, if a correct access token was used in the request.

This issue is separated from #716 and #919 so that testing on those can start earlier, but a release of #919 should not be done without the implementation of this issue.

PR: #937
PR-lib: https://github.com/owncloud/ios-library/pull/162

### PR DESCRIPTION
**Pull request by nasli**
Closes #928 

TASKS
- [X] create new request to refresh token
- [X] retry any network request with a counter
- [X] New OAuth configuration class that manages all parameters needed
- [X] get new refresh token and use it with the retry request, need to move oauth request into the library
- [X] store new credentials 
- [X] retry successfully any normal network request
- [X] retry any upload request with a counter
- [x] retry any download request with a counter

### CODE REVIEW

File: Owncloud iOs Client/Network/Authentication/HandleCredentials.m [hunk 1]
```diff
@@ -0,0 +1,41 @@
+//
```
Comment (davivel) [inline Owncloud iOs Client/Network/Authentication/HandleCredentials.m]: This file is not used in the app, XCode is even ignoring it. Please, remove it.

File: Owncloud iOs Client/Network/Authentication/OauthAuthentication.swift [hunk 2]
```diff
@@ -152,90 +152,83 @@ class OauthAuthentication: NSObject, URLSessionDelegate, URLSessionTaskDelegate
...
-        })
-    }
-    
-    
-    func isAccessTokenDateExpired(_ tokenDate: String?) ->Bool {
-        //TODO: compare dates
-        return true
-    }
-    
+//MOVED TO OC LIBRARY
```
Comment (davivel) [inline Owncloud iOs Client/Network/Authentication/OauthAuthentication.swift]: 🗑

File: Owncloud iOs Client/Network/Downloads/Download.m [hunk 3]
```diff
@@ -94,17 +94,17 @@ - (void) processToDownloadTheFile {
     NSArray *splitedUrl = [[UtilsUrls getFullRemoteServerPath:app.activeUser] componentsSeparatedByString:@"/"];
     NSString *serverUrl = [NSString stringWithFormat:@"%@%@%@",[NSString stringWithFormat:@"%@/%@/%@",[splitedUrl objectAtIndex:0],[splitedUrl objectAtIndex:1],[splitedUrl objectAtIndex:2]], _fileDto.filePath, _fileDto.fileName];
         
-    serverUrl = [serverUrl stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
+    serverUrl = [serverUrl stringByRemovingPercentEncoding];
```
Comment (davivel) [inline Owncloud iOs Client/Network/Downloads/Download.m]: Thanks.

File: Owncloud iOs Client/Network/ReadFolder/DetectListOfFiles.swift [hunk 4]
```diff
@@ -69,11 +84,17 @@ import Foundation
...
             }
+            
+        }) { (_ errorHttp: NSInteger?,_ error: NSError?) in
+            
+            completion(errorHttp, error, nil)
         }
+
     }
     
+    func aaa(){
```
Comment (davivel) [inline Owncloud iOs Client/Network/ReadFolder/DetectListOfFiles.swift]: 🗑

File: Owncloud iOs Client/Tabs/FileTab/FilesViewController.m [hunk 5]
```diff
@@ -372,6 +371,11 @@ - (void)initFilesView {
         NSString *path = _currentRemoteFolder;
         path = [path stringByRemovingPercentEncoding];
         
+        
+        [[AppDelegate sharedOCCommunication] setCredentials:app.activeUser.credDto];
```
Comment (davivel) [inline Owncloud iOs Client/Tabs/FileTab/FilesViewController.m]: Do you remember if this is strictly needed? And why?

File: Owncloud iOs Client/Login/Keychain/OCKeychain.h [hunk 6]
```diff
@@ -15,19 +15,24 @@
...
+#import "UserDto.h"
+#import "Customization.h"
+#import "OCCredentialsStorage.h"
 
 
-@interface OCKeychain : NSObject
+@interface OCKeychain : NSObject <OCCredentialsStorageDelegate>
 
-+(BOOL)setCredentialsOfUser:(UserDto *)user;
++(BOOL)setCredentials:(OCCredentialsDto *)credentials;
```
[the remark this answers is not in the document]
Comment (nasli) [reply Owncloud iOs Client/Login/Keychain/OCKeychain.h]: 👍

File: Owncloud iOs Client/Login/Keychain/OCKeychain.h [hunk 7]
```diff
@@ -48,10 +53,10 @@
  * This method updates all the credentials to use a property to allow to access to them when the passcode system is set.
  * Used in db update 12-13
  */
-+(void)updateAllKeychainsToUseTheLockProperty;
++(BOOL)updateAllOCKeychainItemsToUseTheLockProperty;
 
 
-+(void)updateAllKeychainItemsFromDBVersion21To22ToStoreCredentialsDtoWithBasicAuthenticationAsValue;
++(void)updateAllKeychainItemsFromDBVersion21To22ToStoreCredentialsDtoAsValueAndAuthenticationType;
```
[the remark this answers is not in the document]
Comment (nasli) [reply Owncloud iOs Client/Login/Keychain/OCKeychain.h]: 👍

File: Owncloud iOs Client/Network/Authentication/HandleCredentials.m (hunk 1 above)
Comment (nasli) [reply Owncloud iOs Client/Network/Authentication/HandleCredentials.m]: 👍

File: Owncloud iOs Client/Network/Authentication/OauthAuthentication.swift (hunk 2 above)
Comment (nasli) [reply Owncloud iOs Client/Network/Authentication/OauthAuthentication.swift]: 👍

File: Owncloud iOs Client/Network/Downloads/Download.m (hunk 3 above)
Comment (nasli) [reply Owncloud iOs Client/Network/Downloads/Download.m]: 🐾 💃

File: Owncloud iOs Client/Network/ReadFolder/DetectListOfFiles.swift (hunk 4 above)
Comment (nasli) [reply Owncloud iOs Client/Network/ReadFolder/DetectListOfFiles.swift]: 👍

**Pull request merged**